crates/wac-graph/src/graph.rs

```diff
@@ -0,0 +1,1884 @@
+    },
+    /// The source type does not match the target argument type.
+    #[error("mismatched instantiation argument `{name}`")]
+    ArgumentTypeMismatch {
+        /// The name of the argument.
+        name: String,
+        /// The source of the error.
+        #[source]
+        source: anyhow::Error,
+    },
+    /// The argument has already satisfied.
+    #[error("argument `{name}` has already satisfied")]
+    ArgumentAlreadySatisfied {
+        /// The target node that already has the satisfied argument.
+        node: NodeId,
+        /// The name of the argument that is already satisfied.
+        name: String,
+    },
+    /// The graph contains a cycle.
+    #[error("the graph contains a cycle and cannot be encoded")]
+    GraphContainsCycle {
+        /// The node where the cycle was detected.
+        node: NodeId,
+    },
+    /// The encoding of the graph failed validation.
+    #[error("the encoding of the graph failed validation")]
+    EncodingValidationFailure {
+        /// The source of the validation error.
+        #[source]
+        source: BinaryReaderError,
+    },
+    /// The node cannot be unmarked from export as it is a type definition.
+    #[error("the node cannot be unmarked from export as it is a type definition")]
+    MustExportDefinition,
+    /// An implicit import on an instantiation conflicts with an explicit import node.
+    #[error("an instantiation node implicitly imports an item named `{name}`, but it conflicts with an explicit import node of the same name")]
+    ImplicitImportConfig {
+        /// The existing import node.
+        import: NodeId,
+        /// The instantiation node with the implicit import
+        instantiation: NodeId,
+        /// The name of the conflicting import.
+        name: String,
+    },
+    /// An error occurred while aggregating a type.
+    #[error("failed to aggregate type definition for import `{import}`")]
+    TypeAggregationError {
+        /// The name of the import being aggregated.
+        import: String,
+        /// The type aggregation error.
+        #[source]
+        source: anyhow::Error,
+    },
+}
+
+/// An alias for the result type used by the composition graph.
+pub type GraphResult<T> = std::result::Result<T, Error>;
+
+/// An identifier for a package in a composition graph.
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+pub struct PackageId {
+    /// The index into the graph's packages list.
+    index: usize,
+    /// The generation of the package.
+    ///
+    /// This is used to invalidate identifiers on package removal.
+    generation: usize,
+}
+
+/// An identifier for a node in a composition graph.
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+pub struct NodeId {
+    /// The index into the graph's node list.
+    index: usize,
+    /// The generation of the node.
+    ///
+    /// This is used to invalidate identifiers on node removal.
+    generation: usize,
```

Comment:
Just for my own education: are you using generations here so you don't have to reallocate the entire graph if you remove a node?